amulet/deployer.py

```diff
@@ -330,6 +337,53 @@ def deploy_w_timeout_and_dir(self, timeout, deploy_dir):
             yield
         self.deployed = True
 
+    def action_defined(self, service):
+        """
+        :param service: Name of service to get list of defined actions for.
+
+        Returns the list of actions defined for the service.
+        """
+        if service not in self.services:
+            raise ValueError('Service needs to be added before listing actions.')
+        raw = juju(['action', 'defined', service, '--format', 'json'])
+        return json.loads(raw)
+
+    def action_do(self, unit, action, **kw):
```

Comment:
Not many have, and that's not to say it's "right" by any means, it was a patterned developed several years ago in a vacuum and just happens to persist through to today.

My suggestion is also limited to how actions run currently (only against a given unit) the future, it'll be actions that can be done against units, services, and who knows what else. I'm fine with this method as it is here and would be happy to make the glue code so that my example above works as well.